src/custom/constants/networks.ts

```diff
@@ -28,3 +38,142 @@ export const RPC_URLS: { [key in SupportedChainId]: string } = {
   // [SupportedChainId.CELO_ALFAJORES]: `https://alfajores-forno.celo-testnet.org`,
   [SupportedChainId.GNOSIS_CHAIN]: `https://rpc.gnosis.gateway.fm`,
 }
+
+export const RPC_CONFIG: { [key in SupportedChainId]: { chainId: SupportedChainId; name: string } } = {
+  [SupportedChainId.MAINNET]: {
+    chainId: SupportedChainId.MAINNET,
+    name: 'homestead',
+  },
+  [SupportedChainId.GOERLI]: {
+    chainId: SupportedChainId.GOERLI,
+    name: 'goerli',
+  },
+  [SupportedChainId.GNOSIS_CHAIN]: {
+    chainId: SupportedChainId.GNOSIS_CHAIN,
+    name: 'gnosis',
+  },
+}
+
+/**
+ * We use this class for type inference.
+ *
+ * UrlJsonRpcProvider is an abstract class, therefore we have trouble inferring the type of its constructor.
+ * However, as all of the RPC providers we use extend UrlJsonRpcProvider, we can use this class to infer the type of the constructor, by turning it into a concrete one.
+ */
+class BaseUrlJsonRpcProvider extends UrlJsonRpcProvider {}
+type ProviderClass = Newable<typeof BaseUrlJsonRpcProvider>
+
+/**
+ * UrlJsonRpcProvider instances do not have a getSigner implementation, however we need it.
+ *
+ * As they are all based on JsonRpcProvider, we can use JsonRpcProvider's getSigner implementation.
+ *
+ * Why not use JsonRpcProvider directly? Because it does not have specific API provider optimisations,
+ * and it does not automatically support community resources.
+ *
+ * Why not instantiate and return a JsonRpcSigner? Because it has a constructor guard that prevents it from being instantiated directly.
+ * Intention with this guard is to prevent users from accidentally using a JsonRpcSigner with another provider. Which is not possible here.
+ */
```

Comment:
We've shortly chatted about this part - thought I'd add context here as well.

Current behaviour is that we create a `JsonRpcProvider`, which has the following `getSigner` function ([source](https://github.com/ethers-io/ethers.js/blob/f97b92bbb1bde22fcc44100af78d7f31602863ab/packages/providers/src.ts/json-rpc-provider.ts#L488-L490)).

```
getSigner(addressOrIndex?: string | number): JsonRpcSigner {
    return new JsonRpcSigner(_constructorGuard, this, addressOrIndex);
}
```

This works quite well for us so far. However, all these API based providers are instantiated from abstract class `UrlJsonRpcProvider` which has the following implementation ([source](https://github.com/ethers-io/ethers.js/blob/f97b92bbb1bde22fcc44100af78d7f31602863ab/packages/providers/src.ts/url-json-rpc-provider.ts#L81-L87));

```
getSigner(address?: string): JsonRpcSigner {
        return logger.throwError(
            "API provider does not support signing",
            Logger.errors.UNSUPPORTED_OPERATION,
            { operation: "getSigner" }
        );
    }
```

That is problematic. And it is wrong - verifiably so, because we are able to sign messages after all and all we use is an `InfuraProvider` on our current codebase.

What if we created our own `Signer` then? Well, that also turns out to be problematic because, the available options are;

* Create a `Wallet` instance
  * This is suggested by pretty much all API providers
  * It cannot work for us! `Wallet` instances need your private key
* Create a `VoidSigner`
  * These are readonly signers that cannot sign anything
  * We need to be able to sign things
* Create a `JsonRpcSigner`
  * Not possible, due to the `constructorGuard` above.
  * This is at the heart of why we override through prototype - we don't have a way of working around the `constructorGuard` otherwise (short of patching the package)
  * In discussions it is stated that the intention with this guard is to avoid "create a signer for a provider, and then try to use it with another provider" scenario; which would be a massive footgun. It also does not apply to us here so we should be safe.
* Create a new `Signer` from scratch
  * This would be quite a piece of code, and it would pretty much replicate `JsonRpcSigner`
  * Maintaining / reviewing it I think is not worth avoiding this override

I hope this gives enough context about this part.

Reply:
>We need to be able to sign things

Yes, but only when a wallet is connected. In this situation, we use signer as:
```
const { provider } = useWeb3React()
  const signer = provider.getSigner()
```

Maybe I'm wrong, but it looks like a wallet provider is responsible for signing and we don't need it for JsonRpcSigner, do we?

Reply:
Have you tried to sign an order with using `VoidSigner`?

Reply:
I haven't since it was defined as ([source](https://docs.ethers.org/v5/api/signer/));
> A VoidSigner is a simple Signer which cannot sign.

> Yes, but only when a wallet is connected. In this situation, we use signer as:
>Maybe I'm wrong, but it looks like a wallet provider is responsible for signing and we don't need it for JsonRpcSigner, do we?
Underneath, `web3-react` gets the provider from its connectors. [These providers store providers we supply here internally](https://github.com/Uniswap/web3-react/blob/main/packages/network/src/index.ts#LL74C5-L77C8), as such when you call for `provider.getSigner()` you are actually calling the `getSigner` method of one of the providers created. (Alternatively, if you supply only string links, it creates `JsonRpcProvider` in place. However, that is not super desirable in all cases, e.g. public providers, optimisations etc.)

In most cases? That will be fine. Proposed changes try to rely on your wallet first anyways.

However once that is not the case, you end up hitting `getSigner` method of `UrlJsonRpcProvider` which throws an error.

Reply:
>However once that is not the case, you end up hitting getSigner method of UrlJsonRpcProvider which throws an error.

Is it possible in case when any wallet is connected?

Reply:
In this iteration, that'd be dependent on whether or not `window.ethereum` is present or not. To my understanding, in WalletConnect cases that wouldn't be there, creating that possibility.